Revise Map section of the tutorial (#6944)
sunpy/sunpy: docs/tutorial/maps.rst

Maps
****

> In this section of the tutorial, you will learn about the `Map <sunpy.map.GenericMap>` object, the primary data structure provided by sunpy.

Author: Fair point!

Map objects hold two-dimensional data along with the accompanying metadata.
They can be used with any two-dimensional data array with two spatial axes and FITS-compliant metadata.
As you will see in this tutorial, the primary advantage of using a Map object over just a bare NumPy array is the ability to perform coordinate aware operations on the underlying array, such as rotating the Map to remove the roll angle of an instrument or cropping a Map to a specific field of view.
Additionally, because Map is capable of ingesting data from many different sources, it provides a common interface for any two-dimensional data product.

By the end of this tutorial, you will learn how to create a Map, access the underlying data and metadata, and the basics visualizing a Map.
Additionally, you will learn how to combine multiple Maps into a `~sunpy.map.MapSequence` or a `~sunpy.map.CompositeMap`.
Lastly, you will learn how to create Maps from data sources not supported in sunpy.

> .. note:: In this section and in :ref:`timeseries_guide`, we will use the sample data included with
>           sunpy. These data are primarily useful for demonstration purposes or simple debugging.
>           These files have names like ``sunpy.data.sample.AIA_171_IMAGE`` and
>           ``sunpy.data.sample.RHESSI_IMAGE`` and are automatically downloaded to your computer as
>           you need them. Once downloaded, these sample data files will be paths to their location
>           on your computer.

Reviewer: ```suggestion
.. note:: 
    
   In this section and in :ref:`timeseries_guide`, we will use the sample data included with
          sunpy. 
   These data are primarily useful for demonstration purposes or simple debugging.
   These files have names like ``sunpy.data.sample.AIA_171_IMAGE`` and
          ``sunpy.data.sample.RHESSI_IMAGE`` and are automatically downloaded to your computer as
          you need them.
   Once downloaded, these sample data files will be paths to their location
          on your computer.
```

.. _creating-maps:

Creating Maps
=============

To create a `~sunpy.map.Map` from a sample AIA image,

.. code-block:: python

    >>> import sunpy.map
    >>> import sunpy.data.sample  # doctest: +REMOTE_DATA

>     >>> my_map = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)  # doctest: +REMOTE_DATA

Author: Good point. That is not clear anywhere else.

Maps
****

In this section of the tutorial, you will learn about the `Map <sunpy.map.GenericMap>` object.

> Map objects hold two-dimensional data along with the accompanying metadata.

Reviewer 2: ```suggestion
Map objects hold two-dimensional data along with the accompanying coordinate information and metadata.
```

Reviewer 2: i know the coordinate info is within the metadata, but it doesnt just carry this around it generates the wcs from it, so it may be good to explicitly say "coordinate info" or something along these lines here.

Author: My preference would be to keep the text at the front of the tutorial brief and to show the different components of the metadata in the subsequent sections (which I think this tutorial does). When I wrote "metadata" here, I'm lumping *all* metadata together, including information about the WCS. I'm not sure that distinction between non-WCS metadata and WCS metadata is important at this stage in the tutorial.

They can be used with any two-dimensional data array with two spatial axes and FITS-compliant metadata.

> As you will see in this tutorial, the primary advantage of using a Map object over just a bare NumPy array is the ability to perform coordinate aware operations on the underlying array, such as rotating the Map to remove the roll angle of an instrument or cropping a Map to a specific field of view.

Reviewer 2: ```suggestion
As you will see in this tutorial, the primary advantage of using a Map object over just a bare NumPy array is the ability to perform coordinate aware operations on the underlying array, such as rotating the Map to remove the roll angle of an instrument, cropping a Map to a specific field of view, or reprojecting from one observer to another.
```

Author: This tutorial actually does not (currently) cover anything about reprojection so I don't think it makes sense to include this in the tutorial summary.

           [-128.03072  , -128.03072  , -128.03072  , ..., -128.03072  ,
            -128.03072  , -128.03072  ]], dtype=float32)

This should show a representation of the data as well as some of its associated attributes.
If you are in a Jupyter Notebook, this will show a rich HTML view of the table along with several quick-look plots.
Otherwise, you can use the :meth:`~sunpy.map.GenericMap.quicklook` method to see these quick-look plots.

.. code-block:: python

    >>> my_map.quicklook()  # doctest: +SKIP

.. generate:: html
    :html_border:

    import sunpy.map
    import sunpy.data.sample
    my_map = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)

>     print(my_map._repr_html_())

Reviewer 3: I like this!

Author: I stole it from the Map reference docs!

.. _inspecting-maps:

Inspecting Map Metadata
=======================

The metadata for a Map is exposed via attributes on the Map.
These attributes can be accessed by typing ``my_map.<attribute-name>``.
For example, to access the date of the observation,

.. code-block:: python

    >>> my_map.date  # doctest: +REMOTE_DATA
    <Time object: scale='utc' format='isot' value=2011-06-07T06:33:02.770>

> Notice that this is an `~astropy.time.Time` object which we discussed in the previous section :ref:`time-in-sunpy`.

Reviewer 3: ```suggestion
Notice that this is an `~astropy.time.Time` object which we discussed in the previous :ref:`time-in-sunpy` section of the tutorial.
```

Similarly, we can access the exposure time of the image,

.. code-block:: python

    >>> my_map.exposure_time  # doctest: +REMOTE_DATA
    <Quantity 0.234256 s>

> Notice that this returns an `~astropy.units.Quantity` object which we discussed in the previous section :ref:`units-sunpy`.

Reviewer 3: ```suggestion
Notice that this returns an `~astropy.units.Quantity` object which we discussed in the previous :ref:`units-sunpy` section of the tutorial.
```

The full list of attributes can be found in the reference documentation for `~sunpy.map.GenericMap`.

> These metadata attributes are all derived from the underlying FITS metadata, but are represented as rich Python objects, rather than simple strings or floating point numbers.

Reviewer 3: ```suggestion
These metadata attributes are all derived from the underlying FITS metadata, but are represented as rich Python objects, rather than simple strings or numbers.
```

    >>> my_map.dtype  # doctest: +REMOTE_DATA
    dtype('float32')

> Additional, there are several methods that provide basic summary statistics of the data:

Reviewer 3: ```suggestion
Additionally, there are several methods that provide basic summary statistics of the data:
```

.. code-block:: python

    >>> my_map.center  # doctest: +REMOTE_DATA
    <SkyCoord (Helioprojective: obstime=2011-06-07T06:33:02.770, rsun=696000.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07T06:33:02.770, rsun=696000.0 km): (lon, lat, radius) in (deg, deg, m)
        (-0.00406308, 0.04787238, 1.51846026e+11)>): (Tx, Ty) in arcsec
        (3.22309951, 1.38578135)>
    >>> my_map.bottom_left_coord  # doctest: +REMOTE_DATA
    <SkyCoord (Helioprojective: obstime=2011-06-07T06:33:02.770, rsun=696000.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07T06:33:02.770, rsun=696000.0 km): (lon, lat, radius) in (deg, deg, m)
        (-0.00406308, 0.04787238, 1.51846026e+11)>): (Tx, Ty) in arcsec
        (-1228.76466158, -1224.62447509)>
    >>> my_map.top_right_coord  # doctest: +REMOTE_DATA
    <SkyCoord (Helioprojective: obstime=2011-06-07T06:33:02.770, rsun=696000.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07T06:33:02.770, rsun=696000.0 km): (lon, lat, radius) in (deg, deg, m)
        (-0.00406308, 0.04787238, 1.51846026e+11)>): (Tx, Ty) in arcsec
        (1235.21095899, 1227.39598836)>

> But what if we wanted to know what pixel these physical coordinates correspond to?

Reviewer 3: ```suggestion
But what if we wanted to know what pixel these physical coordinates correspond to, or what physical coordinate corresponds to a particular pixel?
```

Author: Hmm to me, the suggested addition makes this sentence more confusing

Reviewer 2: what about 
```suggestion
But what if we wanted to know what pixel these physical coordinates correspond to? or visa versa
```

Author: Sorry to keep pushing back on this, but I think the original wording works best in the flow of this section. We demonstrate map has these coordinate attributes and then the next natural question is what is the corresponding pixel coordinate? I think presenting things incrementally with limited scope is how a tutorial should be structured. Plus, the converse operation (pixel-to-world) is shown a few lines later so I don't think it would be possible for a user to leave this tutorial thinking WCS does not allow you to go both ways.

> Thankfully, each Map has an associated World Coordinate System, or WCS, which is derived from the underlying metadata and expressed as an `astropy.wcs.WCS` object.

Reviewer 3: ```suggestion
Each Map has an associated World Coordinate System, or WCS, which is derived from the underlying metadata and expressed as an `astropy.wcs.WCS` object.
```

The WCS is accessible as an attribute:

.. code-block:: python

    >>> my_map.wcs  # doctest: +REMOTE_DATA
    WCS Keywords
    <BLANKLINE>
    Number of WCS axes: 2
    CTYPE : 'HPLN-TAN'  'HPLT-TAN'
    CRVAL : 0.00089530541880571  0.00038493926472939
    CRPIX : 512.5  512.5
    PC1_1 PC1_2  : 0.99999706448085  0.0024230207763071
    PC2_1 PC2_2  : -0.0024230207763071  0.99999706448085
    CDELT : 0.00066744222222222  0.00066744222222222
    NAXIS : 1024  1024

> WCS is a fairly complex topic, but all we need to know for now is that the WCS provides the transformation between the pixel coordinates of the image and physical or "world" coordinates.

Reviewer 3: ```suggestion
WCS is a fairly complex topic, but all we need to know for now is that the WCS provides the transformation back and forth between the pixel coordinates of the image and physical or "world" coordinates.
```
To be clear the transform is two way?

Author: I don't think "between" implies a preferred direction. "back and forth between" feels redundant. Furthermore, this bi-directionality is illustrated in the code below.

Reviewer 2: is there somewhere we can point to explain WCS - even a paper or outside sunpy somewhere?

Author: There is an intersphinx link to `astropy.wcs.WCS` above which does include some explanatory docs on WCS. There are the WCS papers, but I would not say those are a gentle introduction to WCS. Maybe adding something like a call out box with some links would be helpful?

In particular, we will only focus on two methods: `~astropy.wcs.WCS.world_to_pixel` and `~astropy.wcs.WCS.pixel_to_world`.
First, let's find the pixel location corresponding to the center of the Map,

.. code-block:: python

    >>> center_pixel = my_map.wcs.world_to_pixel(my_map.center)  # doctest: +REMOTE_DATA
    >>> center_pixel  # doctest: +REMOTE_DATA
    (array(511.5), array(511.5))

> Notice that these coordinates are not necessarily integers.

Reviewer 3: Do we need to explain here, or before, that integers correspond to the lower left corners of pixels, and non-integers are locations within a pixel?

Author: I wondered about that...In the last code snippet in this example, I do state that (0,0) is the center of the lower-left corner pixel. I think this information is important, but I'm worried about it being out of place in a Map tutorial. My preference would be to write a topic guide on pixel coordinate conventions (including diagrams!) and then to refer to that here.

Reviewer 2: hmm I would even suggest defining a new SkyCoord and doing that here instead of using the map center? or show both

Author: Why define a new one? The center/corner attributes are demonstrated above and the user has (presumably) already read the coordinates section of the tutorial which demonstrates how to create a SkyCoord.

    my_map = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)

In the :ref:`creating-maps` section, you learned how to generate a quicklook summary of a Map.
However, the Map object also has a :meth:`~sunpy.map.GenericMap.plot` method that allows for more fine-grained control over how the Map is visualized and is especially useful for generating publication-quality plots.
In this section of the tutorial, you will learn how to build up an increasingly detailed visualization of a Map, including adjusting the colormap and normalization and and overlaying coordinates and contours.

Basic Plotting
--------------

First, let's create a basic plot of our Map, including a colorbar,

.. plot::
    :include-source:
    :context: close-figs

    import matplotlib.pyplot as plt
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_map)
    my_map.plot(axes=ax)
    plt.colorbar()
    plt.show()

.. note::

    We imported `matplotlib.pyplot` in order to create the figure and the axis we plotted on our map onto.
    Under the hood, sunpy uses of `matplotlib` to visualize the image meaning that plots built with sunpy can be further customized using `matplotlib`.

>     **However, for the purposes of this tutorial, you do not need to be familiar with matplotlib.**

Reviewer 3: ```suggestion
    **However, for the purposes of this tutorial, you do not need to be familiar with Matplotlib.**
```

    Fore a series of detailed examples showing how to heavily customize your Map plots, see the :ref:`Plotting section of the Example Gallery <sphx_glr_generated_gallery_plotting>`.

> Note that the title and colormap have been determined based on the accompanying metadata.

Reviewer 3: ```suggestion
Note that the title and colormap have been set by sunpy based on the observing instrument and wavelength.
```

Furthermore, the tick and axes labels have been automatically set based on the coordinate system of the Map.

Looking at the plot above, you likely notice that the resulting image is a bit dim.

> To fix this, we can use the ``clip_interval`` keyword to clip out the dimmest 1% and the brightest 0.5% of pixels.

Reviewer 3: ```suggestion
To fix this, we can use the ``clip_interval`` keyword to automatically adjust the colorbar limits to clip out the dimmest 1% and the brightest 0.5% of pixels.
```

.. plot::
    :include-source:
    :context: close-figs

    import astropy.units as u
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_map)

>     my_map.plot(axes=ax, clip_interval=(1,99.5)*u.percent)

Reviewer 3: ```suggestion
    my_map.plot(axes=ax, clip_interval=(1, 99.5)*u.percent)
```

    my_submap = my_map.submap(roi_bottom_left, top_right=roi_top_right)
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_submap)
    my_submap.plot(axes=ax)
    plt.show()

Additionally, we also may want to combine multiple pixels into a single pixel (a so called "superpixel") to, for example, increase our signal-to-noise ratio.
We can accomplish this with the `~sunpy.map.GenericMap.superpixel` method by specifying how many pixels, in each dimension, we want our new superpixels to contain.
For example, we can combine 4 pixels in each dimension such that our new superpixels contain 16 original pixels.

.. plot::
    :include-source:
    :context: close-figs

    my_super_submap = my_submap.superpixel((5,5)*u.pixel)
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_super_submap)
    my_super_submap.plot(axes=ax)
    plt.show()

.. note::

    Map provides additional methods for manipulating the underlying image data.
    See the reference documentation for `~sunpy.map.GenericMap` for a complete list of available methods as well as the :ref:`Map section of the Example Gallery <sphx_glr_generated_gallery_map>` for more detailed examples.

.. _map-sequences:

> Map Sequences

Reviewer 3: This whole section seems a little light on details and motivation, and I wouldn't be against getting rid of it and starting a new map sequences/compositemap tutorial in future?

Author: Yeah I agree. I removed the CompositeMap section for this reason because I didn't think it added much information. I'm inclined to keep this at least for now only because without it, we essentially have no docs (besides reference) on using MapSequence.

=============

While `~sunpy.map.GenericMap` can only contain a two-dimensional array and metadata corresponding to a single observation, a `~sunpy.map.MapSequence` is comprisded of an ordered list of maps.
By default, the Maps are ordered by their observation date, from earliest to latest date.
A `~sunpy.map.MapSequence` can be created by supplying multiple existing maps:

.. code-block:: python

    >>> another_map = sunpy.map.Map(sunpy.data.sample.EIT_195_IMAGE)  # doctest: +REMOTE_DATA
    >>> map_seq = sunpy.map.Map([my_map, another_map], sequence=True)  # doctest: +REMOTE_DATA

A map sequence can be indexed in the same manner as a list.

> For example, the following should return the same quicklook view as in :ref:`creating-maps`.

Reviewer 3: ```suggestion
For example, the following returns the same information as in :ref:`creating-maps`.
```

    my_map = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)

In the :ref:`creating-maps` section, you learned how to generate a quicklook summary of a Map.
However, the Map object also has a :meth:`~sunpy.map.GenericMap.plot` method that allows for more fine-grained control over how the Map is visualized and is especially useful for generating publication-quality plots.
In this section of the tutorial, you will learn how to build up an increasingly detailed visualization of a Map, including adjusting the colormap and normalization and and overlaying coordinates and contours.

Basic Plotting
--------------

First, let's create a basic plot of our Map, including a colorbar,

.. plot::
    :include-source:
    :context: close-figs

    import matplotlib.pyplot as plt
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_map)
    my_map.plot(axes=ax)
    plt.colorbar()
    plt.show()

.. note::

    We imported `matplotlib.pyplot` in order to create the figure and the axis we plotted on our map onto.
    Under the hood, sunpy uses of `matplotlib` to visualize the image meaning that plots built with sunpy can be further customized using `matplotlib`.

>     **However, for the purposes of this tutorial, you do not need to be familiar with Matplotlib.**

Reviewer 2: I think WCSAxes should be mentioned here somewhere no? just so that people know that its not _exactly_ the same as matplotlib by itself

Author: Fair point. I don't think the main body of the tutorial is a good place, but this call out box is probably a good place to link this information.

Notice that in the plots we've made so far, the ticks on our colorbar are not linearly spaced.
Just like in the case of the colormap, we can use a normalization other than the default by passing a keyword argument to the :meth:`~sunpy.map.GenericMap.plot` method.
For example, we can use a logarithmic normalization instead.

.. plot::
    :include-source:
    :context: close-figs

    import matplotlib.colors
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_map)
    my_map.plot(norm=matplotlib.colors.LogNorm())
    plt.colorbar()
    plt.show()

.. note::

    You can also view or make changes to the default settings through the ``sunpy.map.GenericMap.plot_settings`` dictionary.
    See :ref:`sphx_glr_generated_gallery_plotting_map_editcolormap.py` for an example of of how to change the default plot settings.

.. _wcsaxes-plotting:

Overlaying Contours and Coordinates
-----------------------------------

When plotting images, we often want to highlight certain features or overlay certain data points.
There are several methods attached to Map that make this task easy.

> For example, we can draw contours corresponding to particularly percentages of our data..

Reviewer 4: I would maybe rewrite this: completely:

```suggestion
For example, we can draw contours around the brightest 0.5% percent of pixels in the image:
```

This makes it simple to plot *any* coordinate on top of our Map using the :meth:`~astropy.visualization.wcsaxes.WCSAxes.plot_coord` method.
The following example shows how to plot some points on our Map, including the center coordinate of our Map.

.. plot::
    :include-source:
    :context: close-figs

    coords = SkyCoord(Tx=[100,1000] * u.arcsec, Ty=[100,1000] * u.arcsec, frame=my_map.coordinate_frame)
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_map)
    my_map.plot(axes=ax, clip_interval=(1,99.5)*u.percent)
    ax.plot_coord(coords, 'o')
    ax.plot_coord(my_map.center, 'X')
    plt.show()

.. note::

    Map visualizations can be heavily customized using both `matplotlib` and `astropy.visualization.wcsaxes`.
    See the :ref:`Plotting section of the Example Gallery <sphx_glr_generated_gallery_plotting>` for more detailed examples of how to customize Map visualizations.

.. _cropping-maps:

Cropping Maps and Combining Pixels
==================================

> It analyzing images of the Sun, we often want to choose a smaller portion of the full disk to look at more closely.

Reviewer 2: ```suggestion
In analyzing images of the Sun, we often want to choose a smaller portion of the full disk to look at more closely.
```

Let's use the region of interest we defined above to crop out that portion of our image.

.. plot::
    :include-source:
    :context: close-figs

    my_submap = my_map.submap(roi_bottom_left, top_right=roi_top_right)
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_submap)
    my_submap.plot(axes=ax)
    plt.show()

> Additionally, we also may want to combine multiple pixels into a single pixel (a so called "superpixel") to, for example, increase our signal-to-noise ratio.

Reviewer 2: is the word "combine" here confusing? should it be resample? or something else?

Author: I think combine here is the right word. The image is not being resampled, its being rebinned such that multiple pixels are actually being combined into a single new pixel.

We can accomplish this with the `~sunpy.map.GenericMap.superpixel` method by specifying how many pixels, in each dimension, we want our new superpixels to contain.
For example, we can combine 4 pixels in each dimension such that our new superpixels contain 16 original pixels.

.. plot::
    :include-source:
    :context: close-figs

    my_super_submap = my_submap.superpixel((5,5)*u.pixel)
    fig = plt.figure()
    ax = fig.add_subplot(projection=my_super_submap)
    my_super_submap.plot(axes=ax)
    plt.show()

.. note::

    Map provides additional methods for manipulating the underlying image data.
    See the reference documentation for `~sunpy.map.GenericMap` for a complete list of available methods as well as the :ref:`Map section of the Example Gallery <sphx_glr_generated_gallery_map>` for more detailed examples.

.. _map-sequences:

Map Sequences
=============

> While `~sunpy.map.GenericMap` can only contain a two-dimensional array and metadata corresponding to a single observation, a `~sunpy.map.MapSequence` is comprisded of an ordered list of maps.

Reviewer 2: ```suggestion
While `~sunpy.map.GenericMap` can only contain a two-dimensional array and metadata corresponding to a single observation, a `~sunpy.map.MapSequence` is comprised of an ordered list of maps.
```
